File: backend/app/api/games.py
from __future__ import annotations

from datetime import date
from uuid import UUID

from fastapi import APIRouter, Query

from app.deps import DbSession
from app.schemas.analytics import (
    ClutchResponse,
    DecisionGradesResponse,
    MomentumSwingsResponse,
)
from app.schemas.game import GameDetail, GameRead
from app.schemas.play import PlayRead, PlayWpRead
from app.services.analytics_service import get_clutch_index, get_decision_grades, get_momentum_swings
from app.services.game_service import GameService

router = APIRouter(tags=["games"])


@router.get("/games", response_model=list[GameRead])
async def list_games(
    db: DbSession,
    game_date: date | None = Query(None, alias="date", description="Filter by scheduled date (YYYY-MM-DD)"),
    status: str | None = Query(None, description="Filter by status: scheduled|in_progress|final"),
    season: int | None = Query(None, description="Filter by season year (e.g. 2025)"),
    week: int | None = Query(None, description="Filter by week number (1-18 for regular season)"),
    playoffs: bool = Query(False, description="If true, return only playoff games (week >= 19)"),
) -> list[GameRead]:
    svc = GameService(db)
    games = await svc.list_games(game_date=game_date, status=status, season=season, week=week, playoffs=playoffs)
    return games


@router.get("/games/{game_id}", response_model=GameDetail)
async def get_game(game_id: UUID, db: DbSession) -> GameDetail:
    svc = GameService(db)
    return await svc.get_game(game_id)


@router.get("/games/{game_id}/plays", response_model=list[PlayRead])
async def list_plays(game_id: UUID, db: DbSession) -> list[PlayRead]:
    svc = GameService(db)
    return await svc.list_plays(game_id)


@router.get("/games/{game_id}/wp-history", response_model=list[PlayWpRead])
async def wp_history(game_id: UUID, db: DbSession) -> list[PlayWpRead]:
    """Return all plays for a game with their win-probability and SHAP data.
    Used to seed the WpChart when opening the game detail page after a replay."""
    svc = GameService(db)
    return await svc.list_plays_with_wp(game_id)


@router.get("/games/{game_id}/momentum-swings", response_model=MomentumSwingsResponse)
async def momentum_swings(
    game_id: UUID,
    db: DbSession,
    top: int = Query(3, ge=1, le=10, description="Number of top swings to return"),
) -> MomentumSwingsResponse:
    """Return the top N plays by win-probability swing magnitude."""
    return await get_momentum_swings(db, game_id, top=top)


@router.get("/games/{game_id}/clutch", response_model=ClutchResponse)
async def clutch_index(
    game_id: UUID,
    db: DbSession,
    top: int = Query(5, ge=1, le=20, description="Number of top clutch plays to return"),
) -> ClutchResponse:
    """Return clutch play rankings and team totals for a game."""
    svc = GameService(db)
    game = await svc.get_game(game_id)
    return await get_clutch_index(
        db,
        game_id,
        home_abbr=game.home_team.abbr,
        away_abbr=game.away_team.abbr,
        top_plays=top,
    )


@router.get("/games/{game_id}/decision-grades", response_model=DecisionGradesResponse)
async def decision_grades(
    game_id: UUID,
    db: DbSession,
    top: int = Query(10, ge=1, le=50, description="Number of top decisions to return"),
) -> DecisionGradesResponse:
    """Grade coaching decisions on 4th downs using counterfactual win probability."""
    return await get_decision_grades(db, game_id, top=top)

File: frontend/src/api/games.ts
import client from './client'
import type { Game, GameDetail } from '../types/game'
import type { Play } from '../types/play'
import type { WpDataPoint } from '../types/prediction'
import type { ShapFeature } from '../types/prediction'
import type {
  MomentumSwingsResponse,
  ClutchResponse,
  DecisionGradesResponse,
} from '../types/analytics'

export async function fetchGames(params?: {
  date?: string
  status?: string
  season?: number
  week?: number
  playoffs?: boolean
}): Promise<Game[]> {
  const { data } = await client.get<Game[]>('/games', { params })
  return data
}

export async function fetchGame(id: string): Promise<GameDetail> {
  const { data } = await client.get<GameDetail>(`/games/${id}`)
  return data
}

export async function fetchPlays(gameId: string): Promise<Play[]> {
  const { data } = await client.get<Play[]>(`/games/${gameId}/plays`)
  return data
}

// Shape returned by the backend PlayWpRead schema
interface PlayWpRead {
  id: string
  sequence: number
  home_wp: number
  away_wp: number
  description: string | null
  quarter: number
  game_clock_seconds: number
  score_home: number
  score_away: number
  top_shap: ShapFeature[]
  down: number | null
  yards_to_go: number | null
  yard_line_from_own: number | null
  play_type: string | null
  posteam_abbr: string | null
}

export async function fetchWpHistory(gameId: string): Promise<WpDataPoint[]> {
  const { data } = await client.get<PlayWpRead[]>(`/games/${gameId}/wp-history`)
  return data.map((p) => ({
    sequence: p.sequence,
    playId: p.id,
    homeWp: p.home_wp,
    awayWp: p.away_wp,
    description: p.description,
    quarter: p.quarter,
    gameClock: p.game_clock_seconds,
    scoreHome: p.score_home,
    scoreAway: p.score_away,
    topShap: p.top_shap,
    down: p.down,
    yardsToGo: p.yards_to_go,
    yardLineFromOwn: p.yard_line_from_own,
    playType: p.play_type,
    posteamAbbr: p.posteam_abbr,
  }))
}

export async function fetchMomentumSwings(
  gameId: string,
  top = 3,
): Promise<MomentumSwingsResponse> {
  const { data } = await client.get<MomentumSwingsResponse>(
    `/games/${gameId}/momentum-swings`,
    { params: { top } },
  )
  return data
}

export async function fetchClutchIndex(
  gameId: string,
  top = 5,
): Promise<ClutchResponse> {
  const { data } = await client.get<ClutchResponse>(`/games/${gameId}/clutch`, {
    params: { top },
  })
  return data
}

export async function fetchDecisionGrades(
  gameId: string,
  top = 10,
): Promise<DecisionGradesResponse> {
  const { data } = await client.get<DecisionGradesResponse>(
    `/games/${gameId}/decision-grades`,
    { params: { top } },
  )
  return data
}
